### PR TITLE
Fix joining errors

### PR DESCRIPTION
**Pull request by tbekas**

### CODE REVIEW

File: modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/Joining.scala [hunk 1]
```diff
@@ -234,7 +234,8 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
 
       _ <- localHealthcheck.cancel(registrationRequest.id)
 
-      _ <- clusterStorage.addPeer(peer)
+      addSuccess <- clusterStorage.addPeer(peer)
+      _ <- logger.info(s"Peer ${peer.id.show} already in the cluster").whenA(!addSuccess)
```
Comment (kpudlik) [inline modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/Joining.scala:238]: `.unlessA(addSuccess)`

**Review by kpudlik — APPROVED** [review]

File: modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/Joining.scala (hunk 1 above)
Comment (tbekas) [reply modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/Joining.scala:238]: Fixing

**Review by marcinwadon — APPROVED** [review]

**Comment by marcinwadon** [comment]
once you fix Krystian's comment, looks good to me :+1:

**Review by kpudlik — APPROVED** [review]

File: modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/cluster/storage/ClusterStorage.scala [hunk 2]
```diff
@@ -66,34 +66,37 @@ object ClusterStorage {
       def getPeer(id: PeerId): F[Option[Peer]] =
         peers(id).get
 
-      def addPeer(peer: Peer): F[Unit] =
-        setPeer(peer.id)(peer.some)
+      def addPeer(newPeer: Peer): F[Boolean] =
+        updatePeer(newPeer.id) {
+          case Some(currPeer) =>
+            if (currPeer.session < newPeer.session) newPeer.some
```
Comment (TheMMaciek) [inline modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/cluster/storage/ClusterStorage.scala:72]: May be now inconsistent with changed logic in `validateHandshake` in regards to `<` vs `<=`

File: modules/sdk/src/main/scala/org/tessellation/sdk/http/routes/ClusterRoutes.scala [hunk 3]
```diff
@@ -58,23 +57,16 @@ final case class ClusterRoutes[F[_]: Async](
...
-                case PeerAlreadyConnected(id, _, _, _) => Conflict(s"Peer id=${id} already connected.")
-                case SessionDoesNotExist               => Conflict("Peer does not have an active session.")
-                case CollateralNotSatisfied            => Conflict("Collateral is not satisfied.")
-                case NodeNotInCluster                  => Conflict("Node is not part of the cluster.")
-                case _                                 => InternalServerError("Unknown error.")
-              }
-          )
-
+        joining
+          .joinRequest(collateral.hasCollateral)(joinRequest, joinRequest.registrationRequest.ip)
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply modules/sdk/src/main/scala/org/tessellation/sdk/http/routes/ClusterRoutes.scala:61]: Ok. And why exactly you change that flow loosing the restriction? Previously we required to register exactly from the same address that is being declared (if it worked correctly). I think whether it's a security flaw or not

[the remark this answers is not in the document]
Comment (tbekas) [reply modules/sdk/src/main/scala/org/tessellation/sdk/http/routes/ClusterRoutes.scala:61]: Beause it makes our life easier and it doesn't change much when it comes to security - you could still use IP spoofing if you want to be seen as a different origin.

File: modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/cluster/storage/ClusterStorage.scala (hunk 2 above)
Comment (tbekas) [reply modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/cluster/storage/ClusterStorage.scala:72]: I will address this comment soon. I don't have a good idea on how to change this yet.

Comment (tbekas) [reply modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/cluster/storage/ClusterStorage.scala]: Fixed in the recent changes.

File: modules/sdk/src/main/scala/org/tessellation/sdk/http/routes/ClusterRoutes.scala [hunk 4]
```diff
@@ -58,23 +58,18 @@ final case class ClusterRoutes[F[_]: Async](
...
+          .flatMap(_ => Ok())
+          .recoverWith {
+            case PeerAlreadyJoinedWithNewerSession(id, _, _, _) => Conflict(s"Peer id=${id} already connected.")
```
Comment (marcinwadon) [inline modules/sdk/src/main/scala/org/tessellation/sdk/http/routes/ClusterRoutes.scala:65]: `id.show` and I'd update the message to match the case class name

Comment (tbekas) [reply modules/sdk/src/main/scala/org/tessellation/sdk/http/routes/ClusterRoutes.scala:65]: Fixing

File: modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/cluster/storage/ClusterStorage.scala [hunk 5]
```diff
@@ -66,43 +66,51 @@ object ClusterStorage {
...
         peers(id).get
 
-      def addPeer(peer: Peer): F[Unit] =
-        setPeer(peer.id)(peer.some)
+      def addPeer(peer: Peer): F[Boolean] =
+        updatePeerAndGet(peer.id) {
+          case Some(currPeer) =>
+            if (currPeer.session < peer.session) peer.some
+            else currPeer.some
+          case None => peer.some
```
Comment (kpudlik) [inline modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/cluster/storage/ClusterStorage.scala:74]: So we want `peer` in 2 cases:

- currPeer doesn't exist
- currPeer.session is older than peer.session

If you treat both of these cases as `None` (by using filter) then you can get rid of repeated `peer.some`


```scala
updatePeerAndGet {
  _.filterNot(_.session < peer.session).orElse(peer.some)
}
```

**Review by kpudlik — CHANGES_REQUESTED** [review]

**Review by kpudlik — APPROVED** [review]

**Pull request merged**